### PR TITLE
drivers: add dai interface

### PR DESCRIPTION
**Pull request by juimonen**
Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

### CODE REVIEW

File: include/drivers/dai.h [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+	/** The interface is receiving / transmitting data. */
+	DAI_STATE_RUNNING,
+	/** The interface is clocking but not receiving / transmitting data. */
+	DAI_STATE_PRE_RUNNING,
+	/** The interface is draining its transmit queue. */
+	DAI_STATE_STOPPING,
+	/** TX buffer underrun or RX buffer overrun has occurred. */
+	DAI_STATE_ERROR,
+	/** TBD not used today, here for completeness */
+	DAI_STATE_DRAINING,
```
Comment (lgirdwood) [inline include/drivers/dai.h:295]: Looks like we can combine STOPPING with DRAINING states (based on the comments from the Zephyr I2S code)

File: include/drivers/dai.h [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+	DAI_TRIGGER_DROP,
+	/** @brief Prepare the queues after underrun/overrun error has occurred.
+	 *
+	 * This trigger can be used in ERROR state only and changes the
+	 * interface state to READY.
+	 */
+	DAI_TRIGGER_PREPARE,
+	/** @brief Reset
+	 *
+	 * This trigger frees resources.
```
Comment (lgirdwood) [inline include/drivers/dai.h:353]: back to the initial DAI state on driver init.

File: include/drivers/dai.h [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+	 */
+	DAI_TRIGGER_RESET,
+};
+
+/** DAi properties used for binding with dma and other things */
```
Comment (lgirdwood) [inline include/drivers/dai.h:358]: Should we make this get_properties() return a const structure * that contains all the props instead. Its internal only so no ABI pain.

File: include/drivers/dai.h [hunk 4]
```diff
@@ -0,0 +1,717 @@
...
+	DAI_PROPERTY_FIFO = 0,
+	/* handshake id */
+	DAI_PROPERTY_HANDSHAKE,
+	DAI_PROPERTY_INIT_DELAY,
+};
+
+/** @struct dai_legacy_params
+ * @brief Interface configuration options.
+ *
+ * Copied exactly from existing i2s interface to provide integration path for
```
Comment (lgirdwood) [inline include/drivers/dai.h:370]: Can we name the I2S structure and header we copy from.

File: include/drivers/dai.h [hunk 5]
```diff
@@ -0,0 +1,717 @@
...
+/* TODO 1) array of clocks 2) wallclk naming to be clear */
+struct dai_timestamp_cfg {
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+	int type; /* SSP, DMIC, HDA, etc. */
+	int direction; /* Playback, capture */
+	int index; /* For SSPx to select correct timestamp register */
+	int dma_id; /* DMA instance id */
+	int dma_chan_index; /* Used DMA channel */
+	int dma_chan_count; /* Channels in single DMA */
+};
```
Comment (lgirdwood) [inline include/drivers/dai.h:431]: We could probably use some shorts here to save space.

File: include/drivers/dai.h [hunk 6]
```diff
@@ -0,0 +1,717 @@
...
+	/* simple mem_block interface */
+	int (*read)(const struct device *dev, void **mem_block, size_t *size, int stream_id);
+	int (*write)(const struct device *dev, void *mem_block, size_t size, int stream_id);
+
+	/* timestamp ops */
+	int (*timestamps_config)(const struct device *dev, struct dai_timestamp_cfg *cfg);
+	int (*timestamp_start)(const struct device *dev, struct dai_timestamp_cfg *cfg);
+	int (*timestamp_stop)(const struct device *dev, struct dai_timestamp_cfg *cfg);
+	int (*timestamp_get)(const struct device *dev, struct dai_timestamp_cfg *cfg,
+			     struct dai_timestamp_data *tsd);
```
Comment (lgirdwood) [inline include/drivers/dai.h:468]: We need to comment if mandatory or optional

Comment (lgirdwood) [reply include/drivers/dai.h:468]: i.e. ```/* timestamp ops - optional */```

**Review by lgirdwood — COMMENTED** [review]
Needs more comments. I also thinking we should have a properties struct intead of the enum.....

File: include/drivers/dai.h [hunk 7]
```diff
@@ -0,0 +1,717 @@
...
+/** \brief Types of DAI
+ *
+ * The type of the DAI. This ID type is used to configure bespoke DAI HW
+ * settings.
+ *
+ * DAIs have a lot of physical link feature variability and therefore need
+ * different configuration data to cater for different use cases. We
+ * sometimes may need to pass extra configuration data prior to DAI start.
+ */
+enum dai_type {
```
Comment (andyross) [inline include/drivers/dai.h:245]: Some of these enumerants look like they're runtime-discriminatory ("is this particular device a DMIC or HDA stream?") and some are clearly not (Intel vs. AMD).  I'm wondering if there's an extra level of abstraction needed here, where the SOC level code somewhere defines the enumerants that make sense for the platform being built?

Obviously not a big deal if it's a short list and rarely changes.

File: include/drivers/dai.h (hunk 4 above)
Comment (andyross) [reply include/drivers/dai.h:370]: Alternatively put this in a separate "compatibility" API somewhere, maybe.  IMHO if we're doing something new and better, we should just do it right and present the new API as we want it to look.  Then we can write docs or utilities or shim APIs or whatever for the benefit of older users trying to migrate.

**Review by andyross — COMMENTED** [review]
Some notes on zephyrisms, though note that I'm not a particular authority on driver model patterns.

File: include/drivers/dai.h [hunk 8]
```diff
@@ -0,0 +1,717 @@
...
+ * @brief Public APIs for the DAI (Digital Audio Interface) bus drivers.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_DAI_H_
+#define ZEPHYR_INCLUDE_DRIVERS_DAI_H_
+
+/**
+ * @defgroup dai_interface DAI Interface
+ * @ingroup io_interfaces
+ * @brief DAI (Inter-IC Sound) Interface
```
[the remark this answers is not in the document]
Comment (juimonen) [reply include/drivers/dai.h:33]: ok

File: include/drivers/dai.h [hunk 9]
```diff
@@ -0,0 +1,717 @@
...
+#define DAI_OPT_PINGPONG                    BIT(6)
+
+/** \brief Types of DAI
+ *
+ * The type of the DAI. This ID type is used to configure bespoke DAI HW
+ * settings.
+ *
+ * DAIs have a lot of physical link feature variability and therefore need
+ * different configuration data to cater for different use cases. We
+ * sometimes may need to pass extra configuration data prior to DAI start.
```
[the remark this answers is not in the document]
Comment (juimonen) [reply include/drivers/dai.h:243]: ok

File: include/drivers/dai.h (hunk 1 above)
Comment (juimonen) [reply include/drivers/dai.h:295]: ok

File: include/drivers/dai.h (hunk 2 above)
Comment (juimonen) [reply include/drivers/dai.h:353]: ok

File: include/drivers/dai.h [hunk 10]
```diff
@@ -0,0 +1,717 @@
...
+ * @param dir Stream direction: RX, TX, or both, as defined by DAI_DIR_*.
+ *            The DAI_DIR_BOTH value may not be supported by some drivers.
+ *            For those, the RX and TX streams need to be configured separately.
+ * @param cfg Pointer to the structure containing configuration parameters.
+ *
+ * @retval 0 If successful.
+ * @retval -EINVAL Invalid argument.
+ * @retval -ENOSYS DAI_DIR_BOTH value is not supported.
+ */
+__syscall int dai_z_config_set(const struct device *dev,
```
[the remark this answers is not in the document]
Comment (juimonen) [reply include/drivers/dai.h:496]: ok will remove the syscall stuff for now

File: include/drivers/dai.h [hunk 11]
```diff
@@ -0,0 +1,717 @@
...
+
+	/* generic dai config members */
+	dai_fmt_t format;
+	uint32_t channels;
+	uint32_t frame_clk_freq;
+	uint8_t group_id;
+	uint8_t flags;
+
+	/* optional dai specific params */
+	const void *params;
```
[the remark this answers is not in the document]
Comment (juimonen) [reply include/drivers/dai.h:419]: So the thing is that some of the config structs are huge (for example dmic). Whatever way I define those (typed) it is kind of  ugly, for example the dmic structure has 3 levels of nested structs in arrays. I needed actually to increase some heap size to fit those in. So I would keep some kind of private data pointer even if it might cause some issues.

File: include/drivers/dai.h [hunk 12]
```diff
@@ -0,0 +1,717 @@
...
+	int index; /* For SSPx to select correct timestamp register */
+	int dma_id; /* DMA instance id */
+	int dma_chan_index; /* Used DMA channel */
+	int dma_chan_count; /* Channels in single DMA */
+};
+
+struct dai_timestamp_data {
+	uint64_t walclk; /* Wall clock */
+	uint64_t sample; /* Sample count */
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
```
[the remark this answers is not in the document]
Comment (juimonen) [reply include/drivers/dai.h:436]: Ok, these are the exact copies from sof structs. Anyway, idea here is to get the dai specific audio clock timestamp... we are however using system clock in some cases inside the driver implementation. And the audio clock can be sometimes the same as system clock, I guess, but not always.

So the idea is that when having several different dais running, the application can ask the dai specific time.

for e.g. in sof code
```
       /* NTK was set, get wall clock */
        tsd->walclk = io_reg_read_64(TIMESTAMP_BASE + TS_DMIC_LOCAL_WALCLK);

        /* Sample */
        tsd->sample = io_reg_read_64(TIMESTAMP_BASE + TS_DMIC_LOCAL_SAMPLE);

```
wallclk is the dai spesific clock (but I guess it is some system clock), sample is the frames consumed. So for e.g. in asrc we can detect the interface drift with samples vs. wallclk comparison and vary the rate.

@mwasko I guess you also pointed out some "clock" array needed in some cases, could you elaborate?

File: include/drivers/dai.h [hunk 13]
```diff
@@ -0,0 +1,717 @@
...
+	DAI_PROPERTY_INIT_DELAY,
+};
+
+/** @struct dai_legacy_params
+ * @brief Interface configuration options.
+ *
+ * Copied exactly from existing i2s interface to provide integration path for 
+ * legacy zephyr drivers.
+ *
+ * Memory slab pointed to by the mem_slab field has to be defined and
```
[the remark this answers is not in the document]
Comment (juimonen) [reply include/drivers/dai.h:373]: ok. I will try to remove the whole legacy API behind some api pointer or just remove it completely.

File: include/drivers/dai.h [hunk 14]
```diff
@@ -0,0 +1,717 @@
...
+	uint64_t sample; /* Sample count */
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * For internal use only, skip these in public documentation.
+ */
+__subsystem struct dai_driver_api {
```
[the remark this answers is not in the document]
Comment (juimonen) [reply include/drivers/dai.h:444]: ok I was trying to follow zephyr device conventions, it fits almost exactly to current sof driver model, so I could use it.

File: include/drivers/dai.h (hunk 6 above)
Comment (juimonen) [reply include/drivers/dai.h:468]: ok

File: include/drivers/dai.h (hunk 4 above)
Comment (juimonen) [reply include/drivers/dai.h:370]: ok

File: include/drivers/dai.h [hunk 15]
```diff
@@ -0,0 +1,717 @@
...
+	 * The interface was initialized but is not yet ready to receive /
+	 * transmit data. Call dai_configure() to configure interface and change
+	 * its state to READY.
+	 */
+	DAI_STATE_NOT_READY = 0,
+	/** The interface is ready to receive / transmit data. */
+	DAI_STATE_READY,
+	/** The interface is receiving / transmitting data. */
+	DAI_STATE_RUNNING,
+	/** The interface is clocking but not receiving / transmitting data. */
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply include/drivers/dai.h:288]: @andyross some codecs need the DAI to be clocking with no data for a short time typically (<100ms) so the can lock PLLs and power up analog parts. This state just clocks the interface with no data.

File: include/drivers/dai.h (hunk 7 above)
Comment (lgirdwood) [reply include/drivers/dai.h:245]: Will only change when new DAI IP drivers are added to Zephyr.

File: include/drivers/dai.h [hunk 16]
```diff
@@ -0,0 +1,717 @@
...
+#define DAI_FMT_DATA_FORMAT_MASK            (0x7 << DAI_FMT_DATA_FORMAT_SHIFT)
+
+
+/** @brief Standard DAI Data Format.
+ *
+ * Serial data is transmitted in two's complement with the MSB first. Both
+ * Word Select (WS) and Serial Data (SD) signals are sampled on the rising edge
+ * of the clock signal (SCK). The MSB is always sent one clock period after the
+ * WS changes. Left channel data are sent first indicated by WS = 0, followed
+ * by right channel data indicated by WS = 1.
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply include/drivers/dai.h:69]: We can put this into another ABI file, but these are very useful for developers as this HW interface has to be aligned between the SOC and the Codec/BT/MODEM. Some of the devices (like BT and MODEMs) run closed FW where it's really only possible to figure out the interface with a scope (and compare and align with the SOC here).

File: include/drivers/dai.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply include/drivers/dai.h:373]: Lets completely remove, but keep the DAI (I2S) config macros as they are standard).

File: include/drivers/dai.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply include/drivers/dai.h:419]: We can do a 
```c
union {
    struct intel_dmic *intel_dmic;
    struct intel_ssp *intel_ssp;
    struct nxp_i2s *nxp_i2s;
    /* and so on ... */
};
```
The key point is the data is bespoke configuration that cannot easily be defined by a generic or common API, but is abstracted here with a void * or a union. 

This data comes from topology or ACPI on Intel platforms, but could also come from DT on non Intel platforms (or in non host mode).

File: include/drivers/dai.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (mmaka1) [reply include/drivers/dai.h:55]: Since there are so many integration methods for DAIs, those kind of details may apply to a very small subset while the others are programmed elsewhere, outside the DSP. Therefore not sure if this piece of the documentation should be included in this very common definition.

File: include/drivers/dai.h [hunk 17]
```diff
@@ -0,0 +1,629 @@
...
+ *          .-. .-. .-. .-. .-. .-. .-. .-. .-. .-. .-. .-. .-. .-. .-. .-. .-.
+ *     SCK -' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-
+ *            .-------------------------------.                               .-
+ *     WS  ---'                               '-------------------------------'
+ *         ---.---.---.---.---.---.---.---.---.---.---.---.---.---.---.---.---.-
+ *     SD     |MSB|   |...|   |LSB| x |...| x |MSB|   |...|   |LSB| x |...| x |
+ *         ---'---'---'---'---'---'---'---'---'---'---'---'---'---'---'---'---'-
+ *            | Left channel                  | Right channel                 |
+ */
+#define DAI_FMT_DATA_FORMAT_I2S_LEFT_JUSTIFIED  (3 << DAI_FMT_DATA_FORMAT_SHIFT)
```
Comment (mmaka1) [inline include/drivers/dai.h:132]: Not sure why I2S (which is just one of many DAIs) details are included here, should rather be in a dedicated I2S header.

File: include/drivers/dai.h (hunk 7 above)
Comment (mmaka1) [reply include/drivers/dai.h:231]: Should DAI type be stable? I.e. adding a new Intel DAI in the middle would change the value of subsequent entries - does it matter? Or do you plan to append new DAI types - this would be less readable.

File: include/drivers/dai.h [hunk 18]
```diff
@@ -0,0 +1,629 @@
...
+ * @brief Dai Direction
+ * TODO codec vs soc direction
+ */
+enum dai_dir {
+	/** Receive data */
+	DAI_DIR_RX = 0,
+	/** Transmit data */
+	DAI_DIR_TX,
+	/** Both receive and transmit data */
+	DAI_DIR_BOTH,
```
Comment (mmaka1) [inline include/drivers/dai.h:259]: Easier to validate a single direction capability when using flags (rx=1, tx=2, both=(rx|tx))

File: include/drivers/dai.h [hunk 19]
```diff
@@ -0,0 +1,629 @@
...
+	int (*ts_start)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_stop)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_get)(const struct device *dev, struct dai_ts_cfg *cfg,
+		      struct dai_ts_data *tsd);
+};
+/**
+ * @endcond
+ */
+
+static inline int dai_z_init(const struct device *dev)
```
Comment (mmaka1) [inline include/drivers/dai.h:422]: What does `_z_` mean?

File: include/drivers/dai.h [hunk 20]
```diff
@@ -0,0 +1,629 @@
...
+	const struct dai_driver_api *api =
+		(const struct dai_driver_api *)dev->api;
+
+	if (!api->ts_start)
+		return -EINVAL;
+
+	return api->ts_start(dev, cfg);
+}
+
+static inline int dai_ts_z_stop(const struct device *dev, struct dai_ts_cfg *cfg)
```
Comment (mmaka1) [inline include/drivers/dai.h:598]: Was `dai_z_ts...` previously, but switched to `dai_ts_z_...()` (this function and the next one) - why?

File: include/drivers/dai.h [hunk 21]
```diff
@@ -0,0 +1,629 @@
...
+	int index; /* For SSPx to select correct timestamp register */
+	int dma_id; /* DMA instance id */
+	int dma_chan_index; /* Used DMA channel */
+	int dma_chan_count; /* Channels in single DMA */
+};
+
+struct dai_ts_data {
+	uint64_t walclk; /* Wall clock */
+	uint64_t sample; /* Sample count */
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
```
Comment (mmaka1) [inline include/drivers/dai.h:385]: It is rather _freq_  used in the SOF code, is _rate_ more common in Zephyr?

File: include/drivers/dai.h [hunk 22]
```diff
@@ -0,0 +1,629 @@
...
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * For internal use only, skip these in public documentation.
+ */
+__subsystem struct dai_driver_api {
+	int (*init)(const struct device *dev);
```
Comment (mmaka1) [inline include/drivers/dai.h:394]: Might be more complete to have the documentation of the interface included here, not attached to the wrapper implementation.

File: include/drivers/dai.h [hunk 23]
```diff
@@ -0,0 +1,629 @@
...
+	int (*trigger)(const struct device *dev, enum dai_dir dir, int stream_id,
+		       enum dai_trigger_cmd cmd);
+
+	int (*read)(const struct device *dev, void *buf, size_t *size);
+	int (*write)(const struct device *dev, void *buf, size_t size);
+
+	/* timestamp ops - optional */
+	int (*ts_config)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_start)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_stop)(const struct device *dev, struct dai_ts_cfg *cfg);
```
Comment (mmaka1) [inline include/drivers/dai.h:414]: Could you clarify start and stop flow? What about async operation support w/ callback? (might be useful for hw implementation that may take a while to capture the correlated ts).

File: include/drivers/dai.h (hunk 19 above)
Comment (juimonen) [reply include/drivers/dai.h:422]: just a naming scheme as some of the methods are clashing with sof (like dai_trigger). I'm not sure how to do the namespacing here, it could be that the static methods could be just removed totally.

File: include/drivers/dai.h (hunk 20 above)
Comment (juimonen) [reply include/drivers/dai.h:598]: bug.. need to fix

File: include/drivers/dai.h (hunk 21 above)
Comment (juimonen) [reply include/drivers/dai.h:385]: sorry didn't get this.... these are actually direct copies from sof code. But can very well change to freq if better.

File: include/drivers/dai.h (hunk 22 above)
Comment (juimonen) [reply include/drivers/dai.h:394]: ok... I guess the idea was to make it more pretty, this will come very hard to follow as the api definition will span several screens. Anyway if the wrappers are removed, I need to anyway put some documentation here. At least for me the extensive doxygen type documentation makes it very hard to grasp the big picture.

File: include/drivers/dai.h (hunk 17 above)
Comment (juimonen) [reply include/drivers/dai.h:132]: IMHO very good idea.

File: include/drivers/dai.h (hunk 7 above)
Comment (kv2019i) [reply include/drivers/dai.h:231]: Yup, I agree this looks a bit off. @lgirdwood  proposed "dai_config_ext" and I think that sounds good. I.e. a DAI we can have an enumeration of runtime config objects (e..g filter-coefficient blobs) that drivers can support, or not support. A DAI driver that doesn't need any runtime blobs, or all configuration is static (from DT), can ignore the dai_config_ext. But for DAIs that do need this at runtime, will allow to pass the config via dai_config_ext.

File: include/drivers/dai.h (hunk 18 above)
Comment (kv2019i) [reply include/drivers/dai.h:259]: I think this replaced old interface, so we don't need to stay compatible. It's similar enough, so porting is still quite easy.

File: include/drivers/dai.h (hunk 22 above)
Comment (mmaka1) [reply include/drivers/dai.h:394]: Then I'd just put basic documentation about the op purpose and arguments before the op and move detailed op documentation as paragraph in the parent dai_driver_api structure documentation.

File: include/drivers/dai.h (hunk 3 above)
Comment (juimonen) [reply include/drivers/dai.h]: changed into struct

File: include/drivers/dai.h (hunk 23 above)
Comment (kv2019i) [reply include/drivers/dai.h:414]: @mmaka1  This is basicly documented in detail here:
https://github.com/thesofproject/sof/blob/e5d95d1dd3b336f3c465c137187d820e427c00b4/src/include/sof/audio/component.h#L384
Basicly this allows explicit control of timestamp generation in the DAI. Example implementation in a SOF DAI driver for a I2S/SSP: https://github.com/thesofproject/sof/blob/e5d95d1dd3b336f3c465c137187d820e427c00b4/src/drivers/intel/cavs/timestamp.c#L237

@juimonen  Maybe you can add the SOF component.h documentation for these methods here? For some reason SOF documentation for this is in component header, not the DAI header.

File: include/drivers/dai.h (hunk 18 above)
Comment (juimonen) [reply include/drivers/dai.h:72]: ok

File: include/drivers/dai.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (kv2019i) [reply include/drivers/dai.h:181]: @juimonen How this config relates to the bespoke option we pass later. It seems we are passing two opaque pointers in the same API, is this intentional?

File: include/drivers/dai.h [hunk 24]
```diff
@@ -0,0 +1,480 @@
...
+ * The function can be called in NOT_READY or READY state only. If executed
+ * successfully the function will change the interface state to READY.
+ *
+ * If the function is called with the parameter cfg->frame_clk_freq set to 0
+ * the interface state will be changed to NOT_READY.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param dir Stream direction: RX, TX, or both, as defined by DAI_DIR_*.
+ *            The DAI_DIR_BOTH value may not be supported by some drivers.
+ *            For those, the RX and TX streams need to be configured separately.
```
Comment (kv2019i) [inline include/drivers/dai.h:267]: There's no "dir" parameter anymore.

Comment (juimonen) [reply include/drivers/dai.h]: ok

File: include/drivers/dai.h (hunk 21 above)
Comment (kv2019i) [reply include/drivers/dai.h:206]: I think we can close this in review in Zephyr upstream PR review. We might get guidance on timing data types that align better with Zephyr. To me, "walclk" seems a bit silly acronym.

**Comment by kv2019i** [comment]
Review continued in https://github.com/zephyrproject-rtos/zephyr/pull/42697 . Closing here.
